server/client.go

```diff
@@ -1827,9 +1827,20 @@ func (c *client) markConnAsClosed(reason ClosedState) {
 	// Save off the connection if its a client or leafnode.
 	if c.kind == CLIENT || c.kind == LEAF {
 		if nc := c.nc; nc != nil && c.srv != nil {
+			// Capture the subs and pass them to the go routine. We don't
+			// know for sure if the go routine will execute before or
+			// after closeConnection is done, so we can't have the go
+			// routine set c.subs to nil (to release the connection).
+			var subs []*subscription
+			if len(c.subs) > 0 {
+				subs = make([]*subscription, 0, len(c.subs))
+				for _, sub := range c.subs {
+					subs = append(subs, sub)
+				}
+			}
 			// TODO: May want to send events to single go routine instead
 			// of creating a new go routine for each save.
-			go c.srv.saveClosedClient(c, nc, reason)
+			go c.srv.saveClosedClient(c, nc, subs, reason)
```

Comment:
Could we just pass the c.subs map to saveClosedClient and avoid copy and heap alloc?

That way no additional allocs, but since we pas directly ok if c.subs is nil'd out.

Reply:
Well no, that's what was there originally (meaning saveClosedClient() was using c.subs directly). The issue is that at one point we have to set `c.subs = nil` and we use c.subs in `closeConnection()`, but `markConnAsClosed` (which start the `saveClosedClient` go routine can be invoked either in `closeConnection()` or in some other places when detecting a slow consumer, etc..

We could use some more state to detect when it is safe to set `c.subs = nil`.

(and to answer your comment: if we have c.subs that is nil when using it in saveClosedClient, then we could have closed client connections that do not reflect their real list of subscriptions. I am expecting some tests that would then flap).

Reply:
For instance, we could use a waitgroup/channel to wait for saveClosedClient() to be done using c.subs in closeConnection where I now set the c.subs to nil.

Reply:
I meant still pass it in as an arg to saveClosedConn, so c.subs, that way if it is nil'd out inside client struct won't matter but this would avoid additional allocation plus copy.

Reply:
I would have thought that it would not help for the TestNoRaceConnectionObjectReleased, but seems to be ok. Let me push the modifications and make sure that test is green. I will ping when I confirm it is ok.

Reply:
@derekcollison Ok, Travis is green (at least for push, waiting on PR to complete, but I think it is ok).

Reply:
Ok, Travis all green. Ready for review/merge.

Reply:
Thanks.